File: tools/bench/ops_budget.py
#!/usr/bin/env python3
"""0-9 — "1초에 몇 연산인가"의 실측 (Python 쪽).

"1초 ≈ 10^8 연산"이라는 어림은 연산 하나의 비용이 상수라고 가정한다.
그 가정이 얼마나 틀리는지를 재는 스크립트다.

측정 대상은 산술 명령 하나가 아니라 **알고리즘의 안쪽 루프 한 바퀴**다.
`s += 1` 같은 루프는 C++ 에서 컴파일러가 통째로 지워 버려 비교가 성립하지
않는다(그 현상 자체는 ops_budget.cpp 의 마지막 절에서 따로 보인다).

`ops_budget.cpp` 와 워크로드가 1:1 로 대응한다.

사용법:
    python3.13 tools/bench/ops_budget.py
"""
from __future__ import annotations

import math
import random
import time

REPS = 3


def median(xs):
    return sorted(xs)[len(xs) // 2]


def bench(fn, *args):
    ts = []
    for _ in range(REPS):
        t = time.perf_counter()
        r = fn(*args)
        ts.append(time.perf_counter() - t)
    return median(ts), r


# --- 워크로드: 전부 "루프 한 바퀴 = 연산 1회" ------------------------------

N = 10_000_000


def w_seq_scan(a):
    """연속 접근. 캐시가 최대로 먹히는 경우."""
    s = 0
    for i in range(len(a)):
        s += a[i]
    return s


def w_random_scan(a, idx):
    """무작위 접근. 캐시 미스가 지배하는 경우. 같은 횟수, 같은 연산."""
    s = 0
    for i in range(len(idx)):
        s += a[idx[i]]
    return s


def w_dict_lookup(d, keys):
    s = 0
    for i in range(len(keys)):
        s += d[keys[i]]
    return s


def w_knapsack_inner(dp, w, v, n_updates):
    """0/1 배낭의 안쪽 루프. 실제 DP 갱신 한 번의 비용."""
    cnt = 0
    cap = len(dp) - 1
    while cnt < n_updates:
        for j in range(cap, w - 1, -1):
            cand = dp[j - w] + v
            if cand > dp[j]:
                dp[j] = cand
            cnt += 1
            if cnt >= n_updates:
                break
    return dp[cap]


def _leaf(x):
    return x + 1


def w_func_call(n):
    s = 0
    for _ in range(n):
        s = _leaf(s)
    return s


def main():
    print("=== 알고리즘 안쪽 루프 한 바퀴의 비용 (CPython 3.13) ===")
    print(f"{'워크로드':<26} {'n':>12} {'초':>8} {'ns/op':>9} {'ops/s':>16}")
    rows = []

    def report(name, n, el):
        rows.append((name, n, el))
        print(f"{name:<26} {n:>12,} {el:>8.3f} {el / n * 1e9:>9.1f} {n / el:>16,.0f}")

    a = list(range(N))
    el, _ = bench(w_seq_scan, a)
    report("연속 접근 s += a[i]", N, el)

    random.seed(1)
    idx = [random.randrange(N) for _ in range(N)]
    el, _ = bench(w_random_scan, a, idx)
    report("무작위 접근 s += a[idx[i]]", N, el)
    del idx

    nd = 2_000_000
    d = {i: i for i in range(1 << 20)}
    random.seed(2)
    keys = [random.randrange(1 << 20) for _ in range(nd)]
    el, _ = bench(w_dict_lookup, d, keys)
    report("딕셔너리 조회 d[k]", nd, el)
    del d, keys

    ndp = 5_000_000
    el, _ = bench(w_knapsack_inner, [0] * 10001, 7, 13, ndp)
    report("DP 갱신 dp[j]=max(...)", ndp, el)

    nf = 5_000_000
    el, _ = bench(w_func_call, nf)
    report("함수 호출 f(x)", nf, el)

    print("\n=== 정렬: n log n 한 단위의 비용 ===")
    for n in (100_000, 1_000_000, 5_000_000):
        random.seed(1)
        base = [random.randrange(1 << 30) for _ in range(n)]
        ts = []
        for _ in range(REPS):
            arr = base[:]
            t = time.perf_counter()
            arr.sort()
            ts.append(time.perf_counter() - t)
        el = median(ts)
        units = n * math.log2(n)
        print(f"  n={n:>9,}  sort={el:7.3f}s  n·log2 n={units:>13,.0f}  {units / el:>15,.0f} 단위/s")

    print("\n=== 1초 예산 역산: 이 비용이면 N 이 얼마까지 되는가 ===")
    # 단가를 하나로 정할 수 없다는 것이 이 챕터의 요점이다. 그래서 두 개를 쓴다.
    #   최선 = 연속 접근 (캐시가 최대로 먹히는 안쪽 루프)
    #   최악 = 무작위 접근 (포인터를 따라다니는 안쪽 루프)
    best_ns = rows[0][2] / rows[0][1] * 1e9
    worst_ns = rows[1][2] / rows[1][1] * 1e9
    forms = (
        ("O(n)", lambda n: n),
        ("O(n log n)", lambda n: n * math.log2(max(n, 2))),
        ("O(n^2)", lambda n: n * n),
        ("O(n^3)", lambda n: n * n * n),
        ("O(2^n)", lambda n: 2.0**n),
        ("O(n!)", lambda n: math.gamma(n + 1)),
    )

    def max_n(f, budget):
        lo, hi = 1, 1
        while f(hi) < budget and hi < 10**10:
            hi *= 2
        while lo < hi:
            mid = (lo + hi + 1) // 2
            if f(mid) <= budget:
                lo = mid
            else:
                hi = mid - 1
        return lo

    b_best, b_worst = 1e9 / best_ns, 1e9 / worst_ns
    print(f"  최선 단가 = 연속 접근 {best_ns:.1f} ns  -> 1초에 {b_best:,.0f} 연산")
    print(f"  최악 단가 = 무작위 접근 {worst_ns:.1f} ns  -> 1초에 {b_worst:,.0f} 연산")
    print(f"  {'복잡도':<12} {'최대 N (최선)':>16} {'최대 N (최악)':>16}")
    for label, f in forms:
        print(f"  {label:<12} {max_n(f, b_best):>16,} {max_n(f, b_worst):>16,}")


if __name__ == "__main__":
    main()

File: tools/bench/ops_budget.cpp
// 0-9 — "1초에 몇 연산인가"의 실측 (C++ 쪽).
//
// ops_budget.py 와 워크로드가 1:1 로 대응한다. 루프 몸통은 같은 일을 하고,
// 다른 것은 언어뿐이다.
//
// 누산값을 전부 마지막에 출력에 쓴다. 안 그러면 -O2 가 루프를 통째로 지운다.
// `volatile` 을 쓰지 않는 이유: volatile 은 실제 코테 코드에 없는 메모리
// 접근 제약을 넣어 측정 대상을 바꿔 버린다.
//
// 빌드/실행:
//   g++ -std=c++17 -O2 tools/bench/ops_budget.cpp -o /tmp/ops_budget && /tmp/ops_budget

#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdio>
#include <random>
#include <unordered_map>
#include <vector>
using namespace std;

static const int REPS = 3;
static long long g_sink = 0;

static double median3(double a, double b, double c) {
    double v[3] = {a, b, c};
    sort(v, v + 3);
    return v[1];
}

template <typename F>
static double bench(F f) {
    double t[REPS];
    for (int r = 0; r < REPS; ++r) {
        auto t0 = chrono::steady_clock::now();
        g_sink += f();
        auto t1 = chrono::steady_clock::now();
        t[r] = chrono::duration<double>(t1 - t0).count();
    }
    return median3(t[0], t[1], t[2]);
}

static void report(const char* name, long long n, double el) {
    printf("%-30s %12lld %8.3f %9.2f %16.0f\n", name, n, el, el / n * 1e9, n / el);
}

// 함수 호출 비용을 재려면 인라인을 막아야 한다. 인라인되면 호출이 0회다.
__attribute__((noinline)) static long long leaf(long long x) { return x + 1; }

int main() {
    const long long N = 10000000;

    printf("=== 알고리즘 안쪽 루프 한 바퀴의 비용 (g++ 13 -O2) ===\n");
    printf("%-30s %12s %8s %9s %16s\n", "워크로드", "n", "초", "ns/op", "ops/s");

    vector<long long> a(N);
    for (long long i = 0; i < N; ++i) a[i] = i;

    {  // 연속 접근
        double el = bench([&] {
            long long s = 0;
            for (size_t i = 0; i < a.size(); ++i) s += a[i];
            return s;
        });
        report("연속 접근 s += a[i]", N, el);
    }
    {  // 무작위 접근
        mt19937 rng(1);
        vector<int> idx(N);
        for (long long i = 0; i < N; ++i) idx[i] = (int)(rng() % (unsigned)N);
        double el = bench([&] {
            long long s = 0;
            for (size_t i = 0; i < idx.size(); ++i) s += a[idx[i]];
            return s;
        });
        report("무작위 접근 s += a[idx[i]]", N, el);
    }
    {  // 해시맵 조회
        const long long nd = 2000000;
        unordered_map<int, int> d;
        d.reserve(1 << 21);
        for (int i = 0; i < (1 << 20); ++i) d[i] = i;
        mt19937 rng(2);
        vector<int> keys(nd);
        for (long long i = 0; i < nd; ++i) keys[i] = (int)(rng() % (1u << 20));
        double el = bench([&] {
            long long s = 0;
            for (size_t i = 0; i < keys.size(); ++i) s += d[keys[i]];
            return s;
        });
        report("해시맵 조회 d[k]", nd, el);
    }
    {  // DP 갱신
        const long long ndp = 5000000;
        vector<int> dp(10001, 0);
        int w = 7, v = 13, cap = 10000;
        double el = bench([&] {
            long long cnt = 0;
            while (cnt < ndp) {
                for (int j = cap; j >= w; --j) {
                    int cand = dp[j - w] + v;
                    if (cand > dp[j]) dp[j] = cand;
                    ++cnt;
                    if (cnt >= ndp) break;
                }
            }
            return (long long)dp[cap];
        });
        report("DP 갱신 dp[j]=max(...)", ndp, el);
    }
    {  // 함수 호출
        const long long nf = 5000000;
        double el = bench([&] {
            long long s = 0;
            for (long long i = 0; i < nf; ++i) s = leaf(s);
            return s;
        });
        report("함수 호출 f(x)", nf, el);
    }

    printf("\n=== 정렬: n log n 한 단위의 비용 ===\n");
    for (long long n : {100000LL, 1000000LL, 5000000LL}) {
        mt19937 rng(1);
        vector<int> base(n);
        for (long long i = 0; i < n; ++i) base[i] = (int)(rng() >> 2);
        double t[REPS];
        for (int r = 0; r < REPS; ++r) {
            vector<int> arr = base;
            auto t0 = chrono::steady_clock::now();
            sort(arr.begin(), arr.end());
            auto t1 = chrono::steady_clock::now();
            t[r] = chrono::duration<double>(t1 - t0).count();
            g_sink += arr[0];
        }
        double el = median3(t[0], t[1], t[2]);
        double units = (double)n * log2((double)n);
        printf("  n=%9lld  sort=%7.3fs  n·log2 n=%13.0f  %15.0f 단위/s\n", n, el, units, units / el);
    }

    // --- 컴파일러가 루프를 지우는 현상 ------------------------------------
    // 이 절이 있는 이유: "s += i 를 10^8 번 돌려 보니 0.00초였다"는 측정이
    // 잘못된 결론으로 이어지는 것을 막기 위해서다.
    printf("\n=== -O2 가 지워 버리는 루프 ===\n");
    {
        long long s = 0;
        auto t0 = chrono::steady_clock::now();
        for (long long i = 0; i < 1000000000LL; ++i) s += i;
        auto t1 = chrono::steady_clock::now();
        g_sink += s;
        printf("  s += i 를 10^9 회: %.6fs  (닫힌 형태로 접혀 루프가 없어졌다)\n",
               chrono::duration<double>(t1 - t0).count());
    }

    if (g_sink == 12345678987654321LL) printf("unreachable\n");
    return 0;
}
